### PR TITLE
Memory backed api

### PR DESCRIPTION
**Pull request by joe-iddon**
#### What does this implement/fix? How does it work (high level)? Highlight notable design decisions.

Implements a memory-backed API for the top-level `Arctic()` instance.

Intended usage:

```python
>>> import pandas as pd
>>> ac = Arctic("mem://memory_db") # this name is passed through to the C++ storage, but ignored there
>>> ac.create_library("lib")
...
```

Test definition files which now also verify `mem://`:
- `test_arctic_batch.py` - added via extending `arctic_library`/`arctic_client`
- `test_arctic.py` - added again via `arctic_library`/`arctic_client`
- `test_basic_version_store.py` - parameterised all `lmdb_version_store` fixtures to test the memory adapter too*
- `test_stress_generated.py` - parameterised `lmdb_version_store_tiny_segment` to test the memory adapter too.

\* I did not extend the more specific test fixtures like `lmdb_version_store_tombstone_and_sync_passive`.

These add an extra 15 seconds on top of times around 3 mins for the large test files.

#### Any other comments?

The timings for the stress test (`test_stress_small_row` in `test_stress_generated.py`) are:

Adapter type | Write / sec | Read / sec
--- | --- | ---
lmdb | 97 | 5
mem | 11 | 10

The test is for 1MB of data. The results for the high entropy and low entropy times were within a few seconds of each other. (High entropy has completely random data, low entropy has blocks of repeated rows: aaacccbbb.)

At the moment the name given is unused.

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [x ] Have you updated the relevant docstrings and documentation?
 - [x ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [x ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [x ] Are API changes highlighted in the PR description?
 - [x ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

**Comment by poodlewars** [comment]
Please rebase and squash the commits just before merging this one

File: python/arcticdb/arctic.py [hunk 1]
```diff
@@ -140,6 +147,17 @@ def __init__(self, uri: str, encoding_version: EncodingVersion = DEFAULT_ENCODIN
 
                 Example connection strings are `lmdb:///home/user/my_lmdb` or `lmdb:///home/user/my_lmdb?map_size=2GB`.
 
+
+            In-Memory
+            ----
+                The in-memory connection scheme has the form ``mem://<name>`, e.g. `mem://memory_db`.
```
Comment (poodlewars) [inline python/arcticdb/arctic.py:153]: I wonder what the point of `<name>` is then? For now, all I can it doing is adding confusion. Shall we just have `mem://`?

File: python/tests/integration/arcticdb/test_in_memory.py
Comment (poodlewars) [inline python/tests/integration/arcticdb/test_in_memory.py:1]: What's the point of this one? Doesn't it cover stuff already tested in `test_arctic.py` with the new fixture parameter you added?

File: python/tests/integration/arcticdb/version_store/test_basic_version_store.py
Comment (poodlewars) [inline python/tests/integration/arcticdb/version_store/test_basic_version_store.py:1]: This conflicts with @G-D-Petrov 's PR #789 . https://github.com/man-group/ArcticDB/pull/789/files#diff-820e1d6f588509b49c8841acba4cc4c430a434a90a8fa483130ed6af36076048

Probably just calling your new fixture `basic_store` instead will help make the rebase easier for one of you.

Please talk to Georgi to figure out how / when you two can merge your changes.

**Review by poodlewars — CHANGES_REQUESTED** [review]

**Comment by poodlewars** [comment]
After this is merged we should add some documentation about it to the website docs.arcticdb.io

**Review by willdealtry — CHANGES_REQUESTED** [review]

**Comment by jamesmunro** [comment]
Why slower to write?  This suggests something is wrong.

- [ ] explanation for performance

<html>
<body>


Adapter type | Write / sec | Read / sec
-- | -- | --
lmdb | 97 | 5
mem | 11 | 10


</body>
</html>

File: python/arcticdb/adapters/in_memory_library_adapter.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+
+class InMemoryLibraryAdapter(ArcticLibraryAdapter):
+    """
+    Use in-memory storage.
+
+    Supports any URI that begins with `mem://` - for example, `mem://in_memory_db`.
+    """
+
+    # Only query parameter is cache_size at the moment, but this is not even used.
```
[the remark this answers is not in the document]
Comment (jamesmunro) [reply python/arcticdb/adapters/in_memory_library_adapter.py:40]: - [ ] remove the code to handle the query parameters

File: python/arcticdb/arctic.py (hunk 1 above)
Comment (jamesmunro) [reply python/arcticdb/arctic.py:153]: - [ ] remove name

File: cpp/arcticdb/storage/memory/memory_storage-inl.hpp [hunk 3]
```diff
@@ -31,13 +30,21 @@ namespace arcticdb::storage::memory {
             for (auto &kv : group.values()) {
                 util::variant_match(kv.variant_key(),
                                     [&](const RefKey &key) {
-                                        key_vec[key] = kv.segment();
+                                        auto it = key_vec.find(key);
```
[the remark this answers is not in the document]
Comment (joe-iddon) [reply cpp/arcticdb/storage/memory/memory_storage-inl.hpp:33]: Keeps the `it` variable just in the scope of the if-statement. 👍

**Comment by joe-iddon** [comment]
> Why slower to write? This suggests something is wrong.

Lmdb is presumably slower because it is creating and writing to files whereas the in-memory is just a `std::unordered_map` (really a folly `ConcurrentHashMap`). I'm going to document the in-memory API in another PR after merging this one, and will be investigating lmdb on tmpfs which I can compare timings for against lmdb on a real directory.

I don't think my code being faster suggests anything is wrong.

**Comment by joe-iddon** [comment]
Re: performance tests. Sorry @jamesmunro - the table headers are confusing. They are not writes-per-second, but the time taken, in seconds, to complete the writing test. Good to have some explanations anyway.

File: python/tests/integration/arcticdb/test_in_memory.py
Comment (joe-iddon) [reply python/tests/integration/arcticdb/test_in_memory.py:1]: We discussed that the `arctic_client` in `conftest.py` actually does verify that this works already, but it is good to have an explicit high-level check anyway.

**Comment by joe-iddon** [comment]
Some more timings data (again averaging high entropy and low entropy discrepancies can be ignored):

With 1MB of data:

Adapter type | Write / sec | Read / sec
-- | -- | --
lmdb | 25 | 4
mem | 10 | 4

With 2MB of data:

File: cpp/proto/arcticc/pb2/in_memory_storage.proto [hunk 4]
```diff
@@ -10,7 +10,6 @@ syntax = "proto3";
 package arcticc.pb2.in_memory_storage_pb2;
 
 message Config {
-    uint64 cache_size = 1; // if 0 => 10GB
```
Comment (poodlewars) [inline cpp/proto/arcticc/pb2/in_memory_storage.proto:13]: Scary to change a protobuf in a backwards incompatible way!

In this case probably fine, I doubt anyone has serialized this type of config anywhere. Doesn't seem worth the risk though, suggest just adding a comment to explain that this field is deprecated and unused instead?

File: python/arcticdb/adapters/in_memory_library_adapter.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+from arcticdb.version_store._store import NativeVersionStore
+from arcticdb.adapters.arctic_library_adapter import ArcticLibraryAdapter, set_library_options
+from arcticdb.encoding_version import EncodingVersion
+
+
+class InMemoryLibraryAdapter(ArcticLibraryAdapter):
+    """
+    Use in-memory storage.
+
+    Supports only the URI `"mem://"`.
```
Comment (poodlewars) [inline python/arcticdb/adapters/in_memory_library_adapter.py:25]:
```suggestion
    Only supports the URI `"mem://"`.
```

is a bit more grammatical

File: python/arcticdb/arctic.py [hunk 6]
```diff
@@ -138,6 +145,14 @@ def __init__(self, uri: str, encoding_version: EncodingVersion = DEFAULT_ENCODIN
 
                 Example connection strings are `lmdb:///home/user/my_lmdb` or `lmdb:///home/user/my_lmdb?map_size=2GB`.
 
+
+            In-Memory
+            ----
+                The in-memory connection scheme has the form ``"mem://"`.
+
+                The storage is local to the returned `Arctic` instance.
```
Comment (poodlewars) [inline python/arcticdb/arctic.py:153]: Double backticks around `Arctic`?

File: python/arcticdb/arctic.py [hunk 7]
```diff
@@ -138,6 +145,14 @@ def __init__(self, uri: str, encoding_version: EncodingVersion = DEFAULT_ENCODIN
 
                 Example connection strings are `lmdb:///home/user/my_lmdb` or `lmdb:///home/user/my_lmdb?map_size=2GB`.
 
+
+            In-Memory
+            ----
+                The in-memory connection scheme has the form ``"mem://"`.
```
Comment (poodlewars) [inline python/arcticdb/arctic.py:151]: mismatched backticks

**Review by poodlewars — APPROVED** [review]

File: python/arcticdb/arctic.py (hunk 6 above)
Comment (joe-iddon) [reply python/arcticdb/arctic.py:153]: Thanks. Changes are in the docs PR here: https://github.com/man-group/ArcticDB/pull/895/files#diff-a9727c85029e3b2014647019b25f48df66c2670da1d17f0e2f7e279242f0a8a6R147-R149

File: python/arcticdb/arctic.py (hunk 7 above)
Comment (joe-iddon) [reply python/arcticdb/arctic.py:151]: Changes are in the docs PR here: https://github.com/man-group/ArcticDB/pull/895/files#diff-a9727c85029e3b2014647019b25f48df66c2670da1d17f0e2f7e279242f0a8a6R147-R149

File: python/arcticdb/adapters/in_memory_library_adapter.py (hunk 5 above)
Comment (joe-iddon) [reply python/arcticdb/adapters/in_memory_library_adapter.py:25]: Agreed. I'll make the change in the docs PR.

**Review by willdealtry — APPROVED** [review]

**Comment by joe-iddon** [comment]
For the first merge from master which bought @G-D-Petrov's changes, I had to make these changes to resolve the many merge conflicts: (These are also in the commit message)

  * test_restore_version causes a duplicate SYMBOL_LIST AtomKey. LMDB was silently dealing with this
    by returning a DuplicateKeyException, but in-memory storage was catching this and throwing an error.
    --> Modified in-memory storage to also throw a DuplicateKeyExcpetion like LMDB, so that the
        write_journal method in symbol_list.cpp will catch this and print a debug message silently,
        as is the behaviour of LMDB.
  * fixed lmdb_config being defined in conftest.py for all test types, including things which don't
    know how to ignore it (used Georgi's hack to remove from kwargs before setting the cfg)
  * added in-memory store to the unit tests which tested all backends (in line with Georgi's changes
    as we had agreed in advance for this merge)
  * fixed flaky test_separation_between_libraries test which was comparing lists, not sets
  * added a dynamic_schema unit test method for in-memory store to match the others

**Comment by joe-iddon** [comment]
(Brought in Alex's changes from master to hopefully fix macos pytest fails.)

**Comment by poodlewars** [comment]
Please rebase before merging and

- Squash commits in to one
- Do not include the merge commits

**Review by jamesmunro — APPROVED** [review]

**Pull request merged**